Themes/default/scripts/sceditor.plugins.smf.js

```diff
@@ -0,0 +1,1482 @@
+				if (typeof attrs.height !== "undefined")
+					attribs += ' height="' + attrs.height + '"';
+
+				var contentUrl = smf_scripturl +'?action=dlattach;attach='+ id + ';type=preview;thumb';
+				contentIMG = new Image();
+					contentIMG.src = contentUrl;
+			}
+
+			// If not an image, show a boring ol' link
+			if (typeof contentUrl === "undefined" || contentIMG.getAttribute('width') == 0)
+				return '<a data-type="attach" href="' + smf_scripturl + '?action=dlattach;attach=' + id + ';type=preview;file"' + attribs + '>' + content + '</a>';
+			// Show our purdy li'l picture
+			else
+				return '<img' + attribs + ' src="' + contentUrl + '">';
+		}
+	}
+).set(
+	'email', {
+		allowsEmpty: true,
+		quoteType: sceditor.BBCodeParser.QuoteType.never,
+		tags: {
+			a: {
+				'data-type': ['email']
+			}
+		},
+		format: function (element, content)
+		{
+			if (decodeURI(element.href.substr(7)) === content) {
+				return '[email]' + content + '[/email]';
+			}
+
+			return '[email=' + element.href.substr(7) + ']' + content + '[/email]';
+		},
+		html: function (token, attrs, content)
+		{
+			return '<a data-type="email" href="mailto:' + sceditor.escapeEntities(attrs.defaultattr || content, true) + '">' + content + '</a>';
+		}
+	}
+);
+sceditor.formats.bbcode.set(
+	'url', {
+		allowsEmpty: true,
+		quoteType: sceditor.BBCodeParser.QuoteType.never,
+		format(element, content) {
+			if (element.getAttribute('data-type') != 'url')
+				return content;
+
+			if (decodeURI(element.href).replace(/\/$/, '') === content.replace(/\/$/, '')) {
+				return '[url]' + content + '[/url]';
+			}
+
+			return '[url=' + decodeURI(element.href) + ']' + content + '[/url]';
+		},
+		html(token, {defaultattr}, content) {
+			return '<a data-type="url" href="' + encodeURI(defaultattr || content) + '">' + content + '</a>';
+		}
+	}
+).set(
+	'iurl', {
+		allowsEmpty: true,
+		quoteType: sceditor.BBCodeParser.QuoteType.never,
+		tags: {
+			a: {
+				'data-type': ['iurl']
+			}
+		},
+		format({href}, content) {
+			return '[iurl=' + href + ']' + content + '[/iurl]';
+		},
+		html(token, {defaultattr}, content) {
+			return '<a data-type="iurl" href="' + (defaultattr || content) + '">' + content + '</a>';
+		}
+	})
+.set(
+	'ftp', {
+		allowsEmpty: true,
+		quoteType: sceditor.BBCodeParser.QuoteType.never,
+	}
+);
+
+// This pseudo-BBCode exists only to help the autolinker plugin.
+sceditor.formats.bbcode.set(
+	'nolink', {
+		tags: {
+			span: {
+				'class': 'nolink'
+			},
+		},
+		format: '[nolink]{0}[/nolink]',
+		html: '<span class="nolink">{0}</span>'
+	}
+);
+
+sceditor.formats.bbcode.set(
+	'pre', {
+		tags: {
+			a: {
+				'data-type': ['ftp']
+			}
+		},
+		format({href}, content) {
+			return (href == content ? '[ftp]' : '[ftp=' + href + ']') + content + '[/ftp]';
+		},
+		html(token, {defaultattr}, content) {
+			return '<a data-type="ftp" href="' + (defaultattr || content) + '">' + content + '</a>';
+		}
+	})
```

Comment:
This code seems pretty clearly wrong for the pre BBCode. Also, the pre BBCode is defined again a few lines further down.